### PR TITLE
Security fix for Prototype pollution in fast-toml

### PR DESCRIPTION
**Pull request by arjunshibu**
### 📊 Metadata *

`fast-toml` is vulnerable to `Prototype Pollution`.

### ⚙️ Description *

Prototype Pollution refers to the ability to inject properties into existing JavaScript language construct prototypes, such as objects. JavaScript allows all Object attributes to be altered, including their magical attributes such as `__proto__`, `constructor` and `prototype`. An attacker manipulates these attributes to overwrite, or pollute, a JavaScript application object prototype of the base object by injecting other values. Properties on the Object.prototype are then inherited by all the JavaScript objects through the prototype chain.

### 💻 Technical Description *

Prototype pollution if fixed by not allowing to modify object prototype.

### 🐛 Proof of Concept (PoC) *

1. Create the following PoC and INI files:
```javascript
// poc.js
var fs = require('fs')
var TOML = require('fast-toml')
console.log("Before : " + {}.polluted);
var parsed = TOML.parse(fs.readFileSync('./payload.toml'))
console.log("After : " + {}.polluted);

//payload.toml
[__proto__]
polluted = "Yes! Its Polluted"
```
2. Execute the following commands in terminal:
```bash
npm i fast-toml # Install affected module
node poc.js #  Run the PoC
```
3. Check the Output:
```
Before : undefined
After : Yes! Its Polluted
```

### 🔥 Proof of Fix (PoF) *

Prototype pollution is fixed as seen below.

pof

### 👍 User Acceptance Testing (UAT)

* I've executed unit 
* After fix the functionality is unaffected.

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -348,6 +348,9 @@ function splitElements(raw) {
 */
 function getScope(data, elements=[]) {
 	for (let elt of elements) {
+		if (elt == '__proto__')
```
Comment (Mik317) [inline src/utils.js]: It would be better adding `prototype` and `constructor` to the blacklist :smile:

**Review by Mik317 — CHANGES_REQUESTED** [review]
Still bypassable using the `element` check :cry:

Screenshot from 2020-12-28 18-39-08

Cheers,
Mik

**Comment by mzfr** [comment]
Just want to point out that I think someone else(#1) submitted the fix before this PR.

File: src/utils.js (hunk 1 above)
Comment (arjunshibu) [reply src/utils.js]: @Mik317 I've updated as you suggested. Also used a clean approach with function.
> Still bypassable using the element check 😢

You may have used the vulnerable builld coming with this repo. I didn't commit the built file `dist/node/fast-toml.js` last time. This repo doesn't have package.json file to install dependencies. But there is `rollup.config.js` which is used by `rollup` package to build `fast-toml`. You can build the package by using this command.
```bash
npm i rollup rollup-plugin-terser
npx rollup -c
```

**Comment by huntr-helper** [comment]
Sorry arjunshibu, we enjoyed reviewing your fix but it has not been selected this time. If this bounty has not been closed, please feel free to try again with a new pull request!

We appreciate your effort and look forward to reviewing more of your fixes in the future! 🔨😎

[Come join us on Discord](https://huntr.dev/discord)

**Pull request closed without merging**